conditionally download sauce connect and start automatically

bobholt opened the pull request:
This autostarts sauce connect when running tests on Sauce Labs. If a path to the sauce connect binary is passed in as an argument, that is used, otherwise, it downloads and runs the latest sauce connect binary.

Note: Full Sauce support requires https://github.com/w3c/wptrunner/pull/246 be merged as well.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/wptrunner/253)
<!-- Reviewable:end -->

jgraham commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[wptrunner/environment.py, line 99 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9tODUxvTcXm0Xfazl:-Ki9tODUxvTcXm0Xfazm:bjbuebo) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/environment.py#L99)):*
> ```Python
>         self.cache_manager = multiprocessing.Manager()
>         self.stash = serve.stash.StashServer()
>         self.prerun_process = prerun()
> ```

So this is pretty nitpicky, but I think this has become over-specialized. I wonder if we could make this a list of context managers called e.g. `env_extras` or something and then in `__enter__` have code like

    for cm in self.env_extras:
        cm.__enter__()

and similarly in `__exit__`.

---

*[wptrunner/browsers/sauce.py, line 157 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9ujRnt3k1WZGNgTtw:-Ki9ujRnt3k1WZGNgTtx:bd6aadr) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L157)):*
> ```Python
>         sauce_connect_binary = "./sc-*-linux/bin/sc"
>         sc_url = "https://saucelabs.com/downloads/sc-latest-linux.tar.gz"
>         untar(get(sc_url).raw)
> ```

I feel like this could be simplified a bit; isn't it just

    def get_tar(url, dest):
       resp = requests.get(url)
       resp.raise_for_status
       with tarfile.open(fileobj=CStringIO(.bytes)) as f:
            f.extract_all(path=dest)

I also think that `dest` should be a temporary directory rather than the pwd, which also works better if we turn this into a context manager.

---

*[wptrunner/browsers/sauce.py, line 161 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9voE9OPm2rT3g2j--:-Ki9voE9OPm2rT3g2j-0:benu9tb) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L161)):*
> ```Python
>     sc_process = subprocess.Popen(
>         "%s --user=%s --api-key=%s --no-remove-colliding-tunnels --tunnel-identifier=%s --readyfile=./sauce_is_ready --tunnel-domains web-platform.test *.web-platform.test" % (sauce_connect_binary, sauce_user, sauce_key, sauce_tunnel_id),
>         shell=True
> ```

Why does this need `shell=True`? Can't we pass an array of arguments rather than a string?

---

*[wptrunner/browsers/sauce.py, line 166 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9wAqZwtz6GOpJ65wi:-Ki9wAqZwtz6GOpJ65wj:b35rf4u) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L166)):*
> ```Python
>         time.sleep(5)
> 
>     try:
> ```

We can probably check for `poll()` having a status here explicitly and give a better error message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/253)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/253)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[wptrunner/environment.py, line 99 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9tODUxvTcXm0Xfazl:-KiAJodrskNa5G17Oarp:b-r9c7uu) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/environment.py#L99)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

So this is pretty nitpicky, but I think this has become over-specialized. I wonder if we could make this a list of context managers called e.g. `env_extras` or something and then in `__enter__` have code like

    for cm in self.env_extras:
        cm.__enter__()

and similarly in `__exit__`.
</blockquote></details>

Haven't really dealt with context managers before. I'll give it a go.

---

*[wptrunner/browsers/sauce.py, line 157 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9ujRnt3k1WZGNgTtw:-KiAJWgtrWO8CHWgTc-F:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L157)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

I feel like this could be simplified a bit; isn't it just

    def get_tar(url, dest):
       resp = requests.get(url)
       resp.raise_for_status
       with tarfile.open(fileobj=CStringIO(.bytes)) as f:
            f.extract_all(path=dest)

I also think that `dest` should be a temporary directory rather than the pwd, which also works better if we turn this into a context manager.
</blockquote></details>

Done.

---

*[wptrunner/browsers/sauce.py, line 161 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9voE9OPm2rT3g2j--:-KiA7BetjBkpCf89TH4a:b-1960xs) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L161)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Why does this need `shell=True`? Can't we pass an array of arguments rather than a string?
</blockquote></details>

For some reason it didn't seem to interpret the wildcard character in the binary path in argument mode, but passing the whole command as a string worked. I'll give the arguments another go.

---

*[wptrunner/browsers/sauce.py, line 166 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9wAqZwtz6GOpJ65wi:-KiAJVAShbWKqAA-HGso:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L166)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

We can probably check for `poll()` having a status here explicitly and give a better error message.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/253)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 161 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9voE9OPm2rT3g2j--:-KiAKcjTccpcfhR9Jrmc:b2vxw2e) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L161)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

For some reason it didn't seem to interpret the wildcard character in the binary path in argument mode, but passing the whole command as a string worked. I'll give the arguments another go.
</blockquote></details>

Oh, it probably doesn't. But we can use `glob` to get the actual path before starting the command.  In general using `shell=True` is more error-prone and has security implications if we put user-supplied data in the command line.

---

*[wptrunner/browsers/sauce.py, line 139 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/253#-KiAKHI39uTW4f5kRmIb:-KiAKHI4dVPDdUFFsb5U:bfzkv3n) ([raw file](https://github.com/w3c/wptrunner/blob/4a618addbd507df9d3ad755f619a34c1f71cb0c6/wptrunner/browsers/sauce.py#L139)):*
> ```Python
> 
>     if not sauce_connect_binary:
>         temp_path = tempfile.gettempdir()
> ```

We need to delete this path later (which I why I think that a context manager wrapping all of this makes sense).

---

*[wptrunner/browsers/sauce.py, line 150 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/253#-KiAL-eTAqROu8d8Wu-B:-KiAL-eTAqROu8d8Wu-C:b-rn0x8z) ([raw file](https://github.com/w3c/wptrunner/blob/4a618addbd507df9d3ad755f619a34c1f71cb0c6/wptrunner/browsers/sauce.py#L150)):*
> ```Python
>         time.sleep(5)
> 
>     if sc_process.returncode > 0:
> ```

Do we actually expect the process to have exited here? Otherwise we should perhaps check for  `.poll() is [not] None` here and above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/253)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 161 at r1](https://reviewable.io:443/reviews/w3c/wptrunner/253#-Ki9voE9OPm2rT3g2j--:-KiAW6ld0kcECyryHA1o:b-qwdob3) ([raw file](https://github.com/w3c/wptrunner/blob/db7680d782dc5ff2f0684768d04efbe581d44aef/wptrunner/browsers/sauce.py#L161)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Oh, it probably doesn't. But we can use `glob` to get the actual path before starting the command.  In general using `shell=True` is more error-prone and has security implications if we put user-supplied data in the command line.
</blockquote></details>

I just found the `glob` solution as well. Good to know it's also your recommendation.

---

*[wptrunner/browsers/sauce.py, line 139 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/253#-KiAKHI39uTW4f5kRmIb:-KiAVmQYp2uWdeKPoYQL:blglefc) ([raw file](https://github.com/w3c/wptrunner/blob/4a618addbd507df9d3ad755f619a34c1f71cb0c6/wptrunner/browsers/sauce.py#L139)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

We need to delete this path later (which I why I think that a context manager wrapping all of this makes sense).
</blockquote></details>

I was wondering about the wisdom of relying on the OS's TTL to clear it from `/tmp`. I'll make sure it's taken care of when I work on the context manager piece.

---

*[wptrunner/browsers/sauce.py, line 150 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/253#-KiAL-eTAqROu8d8Wu-B:-KiAWD0gfvP6EFrgrGZN:bel4ikx) ([raw file](https://github.com/w3c/wptrunner/blob/4a618addbd507df9d3ad755f619a34c1f71cb0c6/wptrunner/browsers/sauce.py#L150)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Do we actually expect the process to have exited here? Otherwise we should perhaps check for  `.poll() is [not] None` here and above.
</blockquote></details>

This relies on the fact that if the process is still running, then the `returncode` has been set to `None` by `poll()` and `None < 0`, so the clause is skipped.

I can do `returncode not None and returncode > 0` here if that's clearer. I'll make the `process.poll()` the first part of the `while` check to ensure it always gets called at least once (in the short-circuit case) so that `returncode` is always set.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/253)*
<!-- Sent from Reviewable.io -->

bobholt commented:
@jgraham: I _think_ I've addressed all of your feedback except Context Managers and rebased on `master`. I haven't been successful yet in getting that context managers to work (likely due to my unfamiliarity with the limits and repercussions).

I'm on vacation from now through May 1. You called context managers "nitpicky," so if you want to merge this as-is, I can address first-thing when I get back. If you want to hold this open until then, that is also 100% okay with me. 

In any case, we won't have Sauce results until all 3 PRs are merged (this, web-platform-tests, and prbuildbot) and I can work with @plehegar to get the encrypted sauce key added to web-platform-tests, so that will be on hold until May 1 anyway.

jgraham commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/253)*
<!-- Sent from Reviewable.io -->

jgraham commented:
I think there are a few outstanding issues here, but since you plan to work on it later I'm going to merge this now and then get on with merging the repo into wpt.
